pkg/lib/_internal/dnf5daemon.ts

```diff
@@ -297,28 +340,296 @@ export class Dnf5DaemonManager implements PackageManager {
             }
+
             await call(session, "org.rpm.dnf.v0.Goal", "do_transaction", [{}]);
-        } catch (err) {
-            console.warn("install error", err);
-        } finally {
-            if (session)
-                await close_session(session);
+        });
+    }
+
+    async is_installed(pkgnames: string[]): Promise<boolean> {
+        const uninstalled = new Set(pkgnames);
+
+        await this.with_session(async (session) => {
+            const package_attrs = ["name", "is_installed"];
+
+            const [results] = await call(session, "org.rpm.dnf.v0.rpm.Rpm", "list", [
+                {
+                    package_attrs: { t: 'as', v: package_attrs },
+                    scope: { t: 's', v: "all" },
+                    patterns: { t: 'as', v: pkgnames }
+                }
+            ]) as ListPackage[][];
+
+            for (const pkg of results) {
+                if (pkg.is_installed.v) {
+                    uninstalled.delete(pkg.name.v);
+                }
+            }
+        });
+
+        return uninstalled.size === 0;
+    }
+
+    async install_packages(pkgnames: string[], progress_cb?: ProgressCB): Promise<void> {
+        let last_progress = 0;
+        let total_packages: number;
+
+        function signal_emitted(_path: string, _iface: string, signal: string, args: unknown[]) {
+            switch (signal) {
+            case 'transaction_before_begin':
+                [, total_packages] = args as [string, number];
+                break;
+            case 'transaction_elem_progress': {
+                const [, _last_name, processed,] = args as [string, string, number, number];
+                last_progress = processed / total_packages * 100;
+                break;
+            }
+            }
+
+            if (progress_cb) {
+                progress_cb({
+                    waiting: false,
+                    percentage: last_progress,
+                    cancel: null,
+                });
+            }
+        }
+
+        await this.with_session(async (session) => {
+            await call(session, "org.rpm.dnf.v0.rpm.Rpm", "install", [pkgnames, {}]);
+            const [_transaction_items, result] = await call(session, "org.rpm.dnf.v0.Goal", "resolve", [{}]) as InstallResolveResult;
+            if (result !== 0) {
+                const [problems] = await call(session, "org.rpm.dnf.v0.Goal", "get_transaction_problems", []) as TransactionProblem[][];
+                if (problems.every((p: TransactionProblem) => p.problem.v == GoalProblem.ALREADY_INSTALLED)) {
+                    await call(session, "org.rpm.dnf.v0.Goal", "reset", []);
+                    return;
+                }
+
+                const [problem] = await call(session, "org.rpm.dnf.v0.Goal", "get_transaction_problems_string", []);
+                throw new ResolveError(`Resolving install failed with result=${result}. ${problem}`);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-22611-966030b0-20251202-102504-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)